src/main/scala/Table.scala

```diff
@@ -1,12 +1,497 @@
+
+        // Collect primary keys in primaryKeyColumns
+        val primaryKeyToReturn =
+          collectPrimaryKeyColumns(tableSchemaObject, columns)
+
+        // Collect foreign Keys in foreignKeyColumns
+        val foreignKeyMappings =
+          collectForeignKeyColumns(tableSchemaObject, columns)
+
+        // Collect row titles column
+        var rowTitlesColumns =
+          collectRowTitlesColumns(tableSchemaObject, columns)
+
+        new Table(
+          url = url,
+          id = getId(tableProperties),
+          columns = columns,
+          dialect = getDialect(tableProperties),
+          foreignKeys = foreignKeyMappings, // a new type here?
+          notes = getNotes(tableProperties),
+          primaryKey = primaryKeyToReturn,
+          rowTitleColumns = rowTitlesColumns,
+          schemaId = getMaybeSchemaIdFromTableSchema(tableSchemaObject),
+          suppressOutput = getSuppressOutput(tableProperties),
+          annotations = annotations,
+          warnings = warnings
+        )
+      }
+      case _ =>
+        throw new MetadataError(
+          "Table schema not object, not sure how to process"
+        )
+    }
+  }
+
+  private def ensureColumnsNodeIsArray(
+      tableSchemaObject: ObjectNode
+  ): Option[ErrorMessage] = {
+    if (tableSchemaObject.get("columns").isArray) {
+      None
+    } else {
+      tableSchemaObject.set("columns", JsonNodeFactory.instance.arrayNode())
+      Some(
+        ErrorMessage(
+          "invalid_value",
+          "metadata",
+          "",
+          "",
+          "columns",
+          ""
+        )
+      )
+    }
+  }
+
+  private def validateAndExtractColumnsFromSchema(
+      baseUrl: String,
+      lang: String,
+      inheritedProperties: Map[String, JsonNode],
+      tableSchemaObject: ObjectNode
+  ): (Array[Column], Array[ErrorMessage]) = {
+    var warnings = Array[ErrorMessage]()
+
+    val columnObjects = tableSchemaObject
+      .get("columns")
+      .elements()
+      .asScalaArray
+
+    val columns = columnObjects.zipWithIndex
+      .flatMap(c => {
+        val (col, i) = c
+        col match {
+          case colObj: ObjectNode =>
+            Some(
+              Column.fromJson(i + 1, colObj, baseUrl, lang, inheritedProperties)
+            )
+          case _ => {
+            warnings = warnings :+ ErrorMessage(
+              "invalid_column_description",
+              "metadata",
+              "",
+              "",
+              col.toString,
+              ""
+            )
+            None
+          }
+        }
+      })
+
+    val columnNames = columns.flatMap(c => c.name)
+
+    ensureNoDuplicateColumnNames(columnNames)
+    ensureVirtualColumnsAfterColumns(columns)
+    (columns, warnings)
+  }
+
+  private def ensureNoDuplicateColumnNames(columnNames: Array[String]): Unit = {
+    val duplicateColumnNames = columnNames
+      .groupBy(identity)
+      .filter(grp => grp._2.length > 1)
```

Comment:
Can you change this into one of those special `case` lambdas? Would make it easier to understand.

Reply:
Done